File: Dryft/vr-dryft/Assets/Scripts/Core/ErrorHandler.cs
using UnityEngine;
using System;
using System.Threading.Tasks;
using System.Collections.Generic;

namespace Drift.Core
{
    /// <summary>
    /// Error types for categorization.
    /// </summary>
    public enum ErrorType
    {
        Network,
        Auth,
        Validation,
        Server,
        Timeout,
        Unknown
    }

    /// <summary>
    /// Represents an application error with metadata.
    /// </summary>
    public class AppError : Exception
    {
        public ErrorType Type { get; }
        public string Code { get; }
        public bool Retryable { get; }

        public AppError(ErrorType type, string message, string code = null, bool retryable = true)
            : base(message)
        {
            Type = type;
            Code = code;
            Retryable = retryable;
        }

        public static AppError FromException(Exception ex)
        {
            if (ex == null)
            {
                return new AppError(ErrorType.Unknown, "Unknown error occurred");
            }

            var message = ex.Message.ToLower();

            // Network errors
            if (message.Contains("network") ||
                message.Contains("connection") ||
                message.Contains("socket") ||
                message.Contains("unreachable"))
            {
                return new AppError(ErrorType.Network, "Unable to connect. Please check your internet connection.", retryable: true);
            }

            // Timeout errors
            if (message.Contains("timeout") || message.Contains("timed out"))
            {
                return new AppError(ErrorType.Timeout, "Request timed out. Please try again.", retryable: true);
            }

            // Auth errors
            if (message.Contains("unauthorized") || message.Contains("401") || message.Contains("token"))
            {
                return new AppError(ErrorType.Auth, "Session expired. Please log in again.", "AUTH_EXPIRED", retryable: false);
            }

            // Server errors
            if (message.Contains("500") || message.Contains("server error"))
            {
                return new AppError(ErrorType.Server, "Server error. Please try again later.", retryable: true);
            }

            return new AppError(ErrorType.Unknown, ex.Message, retryable: true);
        }
    }

    /// <summary>
    /// Configuration for retry logic.
    /// </summary>
    public class RetryConfig
    {
        public int MaxRetries { get; set; } = 3;
        public float InitialDelaySeconds { get; set; } = 1f;
        public float MaxDelaySeconds { get; set; } = 10f;
        public float BackoffMultiplier { get; set; } = 2f;
        public HashSet<ErrorType> RetryableTypes { get; set; } = new HashSet<ErrorType>
        {
            ErrorType.Network,
            ErrorType.Timeout,
            ErrorType.Server
        };
    }

    /// <summary>
    /// Handles errors, retry logic, and offline state management.
    /// </summary>
    public class ErrorHandler : MonoBehaviour
    {
        public static ErrorHandler Instance { get; private set; }

        [Header("Settings")]
        [SerializeField] private int maxRetries = 3;
        [SerializeField] private float initialDelaySeconds = 1f;
        [SerializeField] private float maxDelaySeconds = 10f;

        // Events
        public event Action<AppError> OnError;
        public event Action<bool> OnConnectivityChanged;
        public event Action OnSessionExpired;

        // State
        public bool IsOnline { get; private set; } = true;
        private float _lastConnectivityCheck;
        private const float ConnectivityCheckInterval = 5f;

        private void Awake()
        {
            if (Instance != null && Instance != this)
            {
                Destroy(gameObject);
                return;
            }
            Instance = this;
            DontDestroyOnLoad(gameObject);
        }

        private void Update()
        {
            // Periodic connectivity check
            if (Time.time - _lastConnectivityCheck > ConnectivityCheckInterval)
            {
                _lastConnectivityCheck = Time.time;
                CheckConnectivity();
            }
        }

        /// <summary>
        /// Executes an async operation with retry logic.
        /// </summary>
        public async Task<T> WithRetryAsync<T>(
            Func<Task<T>> operation,
            RetryConfig config = null)
        {
            config ??= new RetryConfig
            {
                MaxRetries = maxRetries,
                InitialDelaySeconds = initialDelaySeconds,
                MaxDelaySeconds = maxDelaySeconds
            };

            AppError lastError = null;
            float delay = config.InitialDelaySeconds;

            for (int attempt = 0; attempt <= config.MaxRetries; attempt++)
            {
                try
                {
                    return await operation();
                }
                catch (Exception ex)
                {
                    lastError = ex is AppError appError ? appError : AppError.FromException(ex);

                    // Don't retry non-retryable errors
                    if (!lastError.Retryable || !config.RetryableTypes.Contains(lastError.Type))
                    {
                        HandleError(lastError);
                        throw lastError;
                    }

                    // Don't retry on last attempt
                    if (attempt == config.MaxRetries)
                    {
                        HandleError(lastError);
                        throw lastError;
                    }

                    Debug.Log($"[ErrorHandler] Attempt {attempt + 1}/{config.MaxRetries} failed, retrying in {delay:F1}s...");

                    // Wait before retry with exponential backoff
                    await Task.Delay(TimeSpan.FromSeconds(delay));
                    delay = Mathf.Min(delay * config.BackoffMultiplier, config.MaxDelaySeconds);
                }
            }

            throw lastError ?? new AppError(ErrorType.Unknown, "Operation failed");
        }

        /// <summary>
        /// Handles an error (logging, events, etc.).
        /// </summary>
        public void HandleError(AppError error)
        {
            Debug.LogError($"[ErrorHandler] {error.Type}: {error.Message}");

            OnError?.Invoke(error);

            if (error.Type == ErrorType.Auth)
            {
                OnSessionExpired?.Invoke();
            }

            if (error.Type == ErrorType.Network)
            {
                SetOnlineState(false);
            }

            // Report to analytics
            ReportError(error);
        }

        /// <summary>
        /// Reports an error to analytics/crash reporting.
        /// </summary>
        private void ReportError(AppError error)
        {
            // Log locally for all builds
            Debug.Log($"[ErrorReport] Type: {error.Type}, Code: {error.Code}, Message: {error.Message}");

            // Report to backend analytics if available and online
            if (IsOnline && ApiClient.Instance != null)
            {
                try
                {
                    _ = ApiClient.Instance.PostAsync<object>("/v1/analytics/events", new
                    {
                        event_type = "client_error",
                        properties = new
                        {
                            error_type = error.Type.ToString(),
                            error_code = error.Code ?? "",
                            error_message = error.Message,
                            platform = "vr",
                            unity_version = Application.unityVersion,
                            device_model = SystemInfo.deviceModel
                        }
                    });
                }
                catch
                {
                    // Swallow - error reporting should never crash the app
                }
            }
        }

        /// <summary>
        /// Sets the online state and fires events if changed.
        /// </summary>
        private void SetOnlineState(bool isOnline)
        {
            if (IsOnline != isOnline)
            {
                IsOnline = isOnline;
                OnConnectivityChanged?.Invoke(isOnline);
                Debug.Log($"[ErrorHandler] Connectivity changed: {(isOnline ? "Online" : "Offline")}");
            }
        }

        /// <summary>
        /// Checks network connectivity.
        /// </summary>
        private async void CheckConnectivity()
        {
            bool wasOnline = IsOnline;

            try
            {
                // Check Unity's network reachability
                if (Application.internetReachability == NetworkReachability.NotReachable)
                {
                    SetOnlineState(false);
                    return;
                }

                // Try to reach our API
                var healthCheck = await ApiClient.Instance?.HealthCheckAsync();
                SetOnlineState(healthCheck ?? false);

                // If we just came back online, trigger queue processing
                if (!wasOnline && IsOnline)
                {
                    ProcessOfflineQueue();
                }
            }
            catch
            {
                SetOnlineState(false);
            }
        }

        /// <summary>
        /// Gets a user-friendly error message.
        /// </summary>
        public static string GetUserMessage(AppError error)
        {
            return error.Type switch
            {
                ErrorType.Network => "No internet connection. Please check your network and try again.",
                ErrorType.Auth => "Your session has expired. Please log in again.",
                ErrorType.Timeout => "The request took too long. Please try again.",
                ErrorType.Server => "Something went wrong on our end. Please try again later.",
                ErrorType.Validation => error.Message,
                _ => "Something went wrong. Please try again."
            };
        }

        // =======================================================================
        // Offline Queue
        // =======================================================================

        private Queue<Func<Task>> _offlineQueue = new Queue<Func<Task>>();

        /// <summary>
        /// Enqueues an operation to be executed when back online.
        /// </summary>
        public void EnqueueForOffline(Func<Task> operation)
        {
            _offlineQueue.Enqueue(operation);
            Debug.Log($"[ErrorHandler] Enqueued operation. Queue size: {_offlineQueue.Count}");
        }

        /// <summary>
        /// Processes queued operations when back online.
        /// </summary>
        private async void ProcessOfflineQueue()
        {
            Debug.Log($"[ErrorHandler] Processing offline queue ({_offlineQueue.Count} items)...");

            while (_offlineQueue.Count > 0 && IsOnline)
            {
                var operation = _offlineQueue.Dequeue();
                try
                {
                    await WithRetryAsync(async () =>
                    {
                        await operation();
                        return true;
                    });
                }
                catch (AppError error)
                {
                    if (error.Type == ErrorType.Network)
                    {
                        // Re-enqueue if we went offline
                        _offlineQueue.Enqueue(operation);
                        break;
                    }
                    // Otherwise, discard the failed operation
                    Debug.LogWarning($"[ErrorHandler] Dropped queued operation: {error.Message}");
                }
            }

            Debug.Log($"[ErrorHandler] Queue processing complete. Remaining: {_offlineQueue.Count}");
        }

        /// <summary>
        /// Clears the offline queue.
        /// </summary>
        public void ClearOfflineQueue()
        {
            _offlineQueue.Clear();
        }
    }
}

File: Dryft/mobile/src/utils/errorHandler.ts
import NetInfo, { NetInfoState } from '@react-native-community/netinfo';
import { create } from 'zustand';
import { captureException, addBreadcrumb, setTag } from './sentry';
import {
  ERROR_CODES,
  ERROR_MESSAGES,
  ErrorCode,
  isRetryableError,
} from '../types';

// =============================================================================
// Error Types
// =============================================================================

export const ErrorType = ERROR_CODES;
export type ErrorType = ErrorCode;

export interface AppError {
  type: ErrorCode;
  message: string;
  code?: string;
  status?: number;
  retryable: boolean;
  originalError?: Error;
}

export function classifyError(error: unknown): AppError {
  if (error instanceof Error) {
    const message = error.message.toLowerCase();

    // Network errors
    if (
      message.includes('network') ||
      message.includes('fetch') ||
      message.includes('connection') ||
      message.includes('offline')
    ) {
      return {
        type: ERROR_CODES.NETWORK,
        message: ERROR_MESSAGES.network,
        retryable: isRetryableError(ERROR_CODES.NETWORK),
        originalError: error,
      };
    }

    // Timeout errors
    if (message.includes('timeout') || message.includes('aborted')) {
      return {
        type: ERROR_CODES.TIMEOUT,
        message: ERROR_MESSAGES.timeout,
        retryable: isRetryableError(ERROR_CODES.TIMEOUT),
        originalError: error,
      };
    }

    // Auth errors
    if (
      message.includes('unauthorized') ||
      message.includes('401') ||
      message.includes('token')
    ) {
      return {
        type: ERROR_CODES.AUTH,
        message: ERROR_MESSAGES.auth,
        code: 'AUTH_EXPIRED',
        retryable: isRetryableError(ERROR_CODES.AUTH),
        originalError: error,
      };
    }

    // Server errors
    if (message.includes('500') || message.includes('server')) {
      return {
        type: ERROR_CODES.SERVER,
        message: ERROR_MESSAGES.server,
        retryable: isRetryableError(ERROR_CODES.SERVER),
        originalError: error,
      };
    }
  }

  // Unknown error
  return {
    type: ERROR_CODES.UNKNOWN,
    message: ERROR_MESSAGES.unknown,
    retryable: isRetryableError(ERROR_CODES.UNKNOWN),
    originalError: error instanceof Error ? error : new Error(String(error)),
  };
}

// =============================================================================
// Retry Logic
// =============================================================================

export interface RetryConfig {
  maxRetries: number;
  initialDelayMs: number;
  maxDelayMs: number;
  backoffMultiplier: number;
  retryableErrors: ErrorType[];
}

export const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 3,
  initialDelayMs: 1000,
  maxDelayMs: 10000,
  backoffMultiplier: 2,
  retryableErrors: [ERROR_CODES.NETWORK, ERROR_CODES.TIMEOUT, ERROR_CODES.SERVER],
};

export async function withRetry<T>(
  operation: () => Promise<T>,
  config: Partial<RetryConfig> = {}
): Promise<T> {
  const cfg = { ...DEFAULT_RETRY_CONFIG, ...config };
  let lastError: AppError | null = null;
  let delay = cfg.initialDelayMs;

  for (let attempt = 0; attempt <= cfg.maxRetries; attempt++) {
    try {
      return await operation();
    } catch (error) {
      lastError = classifyError(error);

      // Don't retry non-retryable errors
      if (!lastError.retryable || !cfg.retryableErrors.includes(lastError.type)) {
        throw lastError;
      }

      // Don't retry on last attempt
      if (attempt === cfg.maxRetries) {
        throw lastError;
      }

      console.log(
        `[Retry] Attempt ${attempt + 1}/${cfg.maxRetries} failed, retrying in ${delay}ms...`
      );

      // Wait before retry with exponential backoff
      await sleep(delay);
      delay = Math.min(delay * cfg.backoffMultiplier, cfg.maxDelayMs);
    }
  }

  throw lastError;
}

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// =============================================================================
// Network State Management
// =============================================================================

interface NetworkState {
  isConnected: boolean;
  isInternetReachable: boolean | null;
  connectionType: string | null;

  // Actions
  updateState: (state: NetInfoState) => void;
}

export const useNetworkStore = create<NetworkState>((set) => ({
  isConnected: true,
  isInternetReachable: true,
  connectionType: null,

  updateState: (state: NetInfoState) => {
    set({
      isConnected: state.isConnected ?? true,
      isInternetReachable: state.isInternetReachable,
      connectionType: state.type,
    });
  },
}));

// Initialize network listener
let unsubscribe: (() => void) | null = null;

export function initNetworkListener(): void {
  if (unsubscribe) return;

  unsubscribe = NetInfo.addEventListener((state) => {
    useNetworkStore.getState().updateState(state);

    if (!state.isConnected) {
      console.log('[Network] Device went offline');
    } else if (state.isConnected && state.isInternetReachable === false) {
      console.log('[Network] Connected but no internet access');
    }
  });
}

export function stopNetworkListener(): void {
  if (unsubscribe) {
    unsubscribe();
    unsubscribe = null;
  }
}

// =============================================================================
// Offline Queue
// =============================================================================

interface QueuedOperation {
  id: string;
  operation: () => Promise<unknown>;
  timestamp: number;
  retries: number;
}

interface OfflineQueueState {
  queue: QueuedOperation[];
  isProcessing: boolean;

  // Actions
  enqueue: (operation: () => Promise<unknown>) => string;
  dequeue: (id: string) => void;
  processQueue: () => Promise<void>;
  clearQueue: () => void;
}

export const useOfflineQueueStore = create<OfflineQueueState>((set, get) => ({
  queue: [],
  isProcessing: false,

  enqueue: (operation: () => Promise<unknown>) => {
    const id = `op_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    const queuedOp: QueuedOperation = {
      id,
      operation,
      timestamp: Date.now(),
      retries: 0,
    };

    set((state) => ({ queue: [...state.queue, queuedOp] }));
    console.log(`[OfflineQueue] Enqueued operation ${id}`);

    // Try to process if we're online
    const networkState = useNetworkStore.getState();
    if (networkState.isConnected && networkState.isInternetReachable) {
      get().processQueue();
    }

    return id;
  },

  dequeue: (id: string) => {
    set((state) => ({
      queue: state.queue.filter((op) => op.id !== id),
    }));
  },

  processQueue: async () => {
    const { queue, isProcessing } = get();
    if (isProcessing || queue.length === 0) return;

    const networkState = useNetworkStore.getState();
    if (!networkState.isConnected) return;

    set({ isProcessing: true });

    for (const op of queue) {
      try {
        await withRetry(op.operation, { maxRetries: 2 });
        get().dequeue(op.id);
        console.log(`[OfflineQueue] Processed operation ${op.id}`);
      } catch (error) {
        const appError = classifyError(error);

        if (appError.type === ErrorType.NETWORK) {
          // Stop processing if we lost connection
          console.log('[OfflineQueue] Lost connection, stopping queue processing');
          break;
        }

        // Remove non-retryable operations
        if (!appError.retryable || op.retries >= 3) {
          get().dequeue(op.id);
          console.log(`[OfflineQueue] Removed failed operation ${op.id}`);
        } else {
          // Increment retry count
          set((state) => ({
            queue: state.queue.map((q) =>
              q.id === op.id ? { ...q, retries: q.retries + 1 } : q
            ),
          }));
        }
      }
    }

    set({ isProcessing: false });
  },

  clearQueue: () => {
    set({ queue: [], isProcessing: false });
  },
}));

// Auto-process queue when coming online
useNetworkStore.subscribe((state, prevState) => {
  if (state.isConnected && !prevState.isConnected) {
    console.log('[Network] Back online, processing queued operations');
    useOfflineQueueStore.getState().processQueue();
  }
});

// =============================================================================
// User-Facing Error Messages
// =============================================================================

export function getUserMessage(error: AppError): string {
  if (error.type === ERROR_CODES.VALIDATION && error.message) {
    return error.message;
  }
  return ERROR_MESSAGES[error.type] ?? ERROR_MESSAGES.unknown;
}

// =============================================================================
// Error Reporting (placeholder for analytics)
// =============================================================================

export function reportError(error: AppError, context?: Record<string, unknown>): void {
  console.error('[ErrorReport]', {
    type: error.type,
    message: error.message,
    code: error.code,
    status: error.status,
    context,
    timestamp: new Date().toISOString(),
  });

  // Send to Sentry
  setTag('error_type', error.type);
  setTag('error_code', error.code ?? error.type);

  addBreadcrumb(error.message, 'error', {
    type: error.type,
    retryable: error.retryable,
    ...context,
  });

  if (error.originalError) {
    captureException(error.originalError, {
      errorType: error.type,
      errorCode: error.code,
      retryable: error.retryable,
      ...context,
    });
  }
}
